Integrate `kw_bd` to the project and add migration script

davidbtadokoro opened the pull request:
This PR is a continuation of the work made by @kwy95 and @magalilemes in [**src: db: introduce db model and migration script**](https://github.com/kworkflow/kworkflow/pull/531) (thanks for the excellent work!). Its purpose is to integrate `kw_bd`, the implementation of a real database management system that uses SQLite underneath, into the project. In other words, wherever metadata files are being used to simulate a `DB` or wherever `DB`'s functionalities are advantageous, this integration should happen. Also, this PR adds a script to migrate old metadata files into the `DB`.

**[TODOs]** At the moment this PR is just a draft as there are several missing parts of it:
- ~~Update `kernel-config-manager` test suite (integration is done)~~
- ~~FInish integration with `build` and `deploy` (statistics)~~
- ~~Integrate with `pomodoro`~~
- ~~Integrate with `report`~~
- ~~Fix migration script (migrate the same legacy user data only once)~~

**[TESTING THIS PR]**
If you wish to test this PR (thanks a lot!), below are the steps and the assertions you should make. The TLDR is that just by pulling this PR, installing the PR version of kw and asserting the migration is consistent using `kw report` and `kw kernel-config-manager` is of great help! The steps below are a detailed script to fully test the PR (besides automated tests).

1. It is **REALLY IMPORTANT** to make the backup of your kw personal data. Either use `kw backup <path_to_backup>` or `cp -r <kw_data_dir_path> <path_to_backup>`
2. Have the `unstable` branch updated with `git pull origin unstable`
3. Checkout to the `unstable` branch with `git checkout unstable`
4. Pull this PR locally with `git pull origin pull/836/head:<name_of_branch>`
5. Checkout to the local PR branch with `git switch <name_of_branch>`
6. Install this version of kw with `./setup.sh --install`. You should see a warning like `Creating database: <kw_data_dir_path>/kw.db` and no errors
7. Assert that the file `<kw_data_dir_path>/kw.db` has been created
8. Assert that in your `<kw_data_dir_path>`, the directories `pomodoro` and `statistics` are renamed to `legacy_pomodoro` and `legacy_statistics`, respectively. For the kernel configs, assert that in your `<kw_data_dir_path>/configs`, the directories `configs` and `metadata` are renamed to `legacy_configs` and `legacy_metadata` and that the actual kernel config files (the kernel `.config`s) are in `<kw_data_dir_path>/configs`
9. Run `kw report <period of time>` of a period of time that you can assert the output
10. Experiment with `kw build`, `kw deploy`, `kw deploy --list`, `kw deploy --uninstall` or `kw deploy --modules-deploy` and assert through `kw report --statistics` their addition
11. Experiment with `kw pomodoro --set-timer <timebox>` and assert through `kw report --pomodoro` their addition. Also use `kw pomodoro --check-timer` to see if the active timeboxes are displayed correctly
12. Experiment with `kw kernel-config-manager` and assert that the feature works as intended
13. Reinstall this version of kw with `./setup.sh --install`. You should **NOT** see a warning like `Creating database: <kw_data_dir_path>/kw.db` (also, no errors should happen)
14. Assert that `<kw_data_dir_path>` hasn't changed and that there is no duplicated data. You could use `sqlite3 <kw_data_dir_path>/kw.db` and check that the tables are consistent (e.g. `SELECT * FROM statistics_report;`) or juts use `kw report`
15. If any corruption of your data has happened, restore your backup (recommend deleting the contents of `<kw_data_dir_path>` and running `kw backup --restore <path_to_backup> --force`)
16. THANK YOU SO MUCH FOR TESTING THIS PR!

kwy95 committed:
src: db: introduce db model and migration script

Introduces the file that builds the model to be used by kw to store all
data related to statistics options, pomodoro sessions and kernel config
files.

Adds a script to handle the data migration from the metadata files in
`KW_DATA_DIR` to the newly introduced database. Current migrated data
are: statistics and pomodoro reports, and kernel configs. The
directories holding the metadata are renamed as such:
- `KW_DATA_DIR`/statistics->`KW_DATA_DIR`/legacy_statistics
- `KW_DATA_DIR`/pomodoro->`KW_DATA_DIR`/legacy_pomodoro
- `KW_DATA_DIR`/configs/configs->`KW_DATA_DIR`/configs/legacy_configs
- `KW_DATA_DIR`/configs/metadata->`KW_DATA_DIR`/configs/legacy_metadata

Thanks: Magali Lemes <magalilemes00@gmail.com> for the help modelling
  the database.
Signed-off-by: Rubens Gomes Neto <rubens.gomes.neto@alumni.usp.br>
Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

kwy95 committed:
src: kwlib: Integrate statistics_manager with kw_db

Much of the statistics is done via the `statistics_manager` function
defined inside `kwlib.sh`. Those statistics are represented using
metadata files.

This commit integrates the function with kw_db and makes the necessary
changes in `deploy.sh` and in `build.sh`, which are the callers.

Signed-off-by: Rubens Gomes Neto <rubens.gomes.neto@usp.br>
Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

kwy95 committed:
src: kw_db: improve parsing of NULL

Prevent `format_values_db` from adding single quotes to NULL values.

Signed-off-by: Rubens Gomes Neto <rubens.gomes.neto@usp.br>
Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

kwy95 committed:
src: kernel_config_manager: Integrate with kw_db

kw kernel-config-manager uses metadata files to manage the database
management of kernel configuration files.

This commit changes this by integrating kernel-config-manager with kw_db
(that uses sqlite underneath). This is a more scalable approach and
easier to maintain as some logic of data management is shifted to kw_db.
The actual .config file is still stored in the user filesystem and just
the metadata files were transferred to the database.

Signed-off-by: Rubens Gomes Neto <rubens.gomes.neto@usp.br>
Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

Commit:
src: pomodoro: Update codestyle and refactor structure

The pomodoro.sh file doesn't follow the current project codestyle, like:
- Having a standardized parser
- Parser named 'parse_<feature-name>'
- String concatenation without brackets
- No parser error handling

Also, the structure can be improved as some refactorings can be made,
like method extractions, positioning of functions to reduce file
scrolling and refactoring of options.

This commit fixes those resulting in a more clean code. The actual size
of the file grew, but this is due to the extraction of methods and their
documentation. The 'kw pomodoro --tag' no longer shows all registered
tags, instead this functionality is provided by the '--show-tags/-s'
option. Also, the '--list/-l' option was renamed to '--check-time/-c' to
be more indicative of its behavior.

Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

Commit:
tests: kernel_config_manager_test: Update codestyle and refactor structure

The kw kernel-config-manager test suite is outdated in terms of
codestyle. Some ShellCheck violations like not double quoting to prevent
globbing and word splitting, and more subtle violations like not using
curly braces in string concatenations are present.

Also, some refactorings can be made, like:
- Encapsulate global variable declarations inside `oneTimeSetUp` (those
  were loose at the start of the file);
- Encapsulate changing dirs in and out of the `SHUNIT_TMP_DIR` at the
  beginning and ending of most tests inside `setUp` and `tearDown`,
  respectively;
- Fix the order of arguments in `assert_equals_helper` calls. The second
  to last argument should be <expected_value> and the last should be
  <actual_value>. This order was inverted in most of the file;
- Use the `is_safe_path_to_remove` function every time an `rm` or
  `rm -rf` is used for safekeeping;
- Fix variable named `ret`. It was used in most tests for the same
  purpose as the `output` variable used throughout the project (i.e. to
  capture the output of a command to assert something about it). Besides
  the name not following the project codestyle, it also was wrongly
  used, in some cases;
- Change `assertTrue` calls by `assert_equals_helper` where convenient.
  The first function was used to assert if two values were exactly equal
  in some cases. For those, it is better to use the second function as
  it shows the expected and actual values with colors while the first
  just prints a message.

This commit addresses the points described above to make the suite more
maintainable and also adds some TODO comments where some further changes
can be done that escape the scope of this commit.

Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

davidbtadokoro commented:
**[UPDATE]**
This force push adds a commit that greatly refactors the `kernel-config-manager` test suite (in the `tests/kernel_config_manager_test.sh` file). Besides making the code more clean and updated with the project code style, the objective was to make it easier to alter it for the `kw_db` integration (which is the next step in this PR).

davidbtadokoro commented:
**[UPDATE]**
These last couple force pushes made more codestyle and refactoring updates to the `kw kernel-config-manager` test suite (in commit `tests: kernel_config_manager_test: Update codestyle and refactor structure`) and updated the test suite for the `kw_db` integration.

davidbtadokoro commented:
**[UPDATE]**
This force push adds the integration with `build` and `deploy`. But there are some fixes needed to be done before consider it finished. The commit [src: kwlib: Integrate with kw_db](https://github.com/kworkflow/kworkflow/pull/836/commits/de8d14d91b8d0ba5d2b8a3ceb6883f0bfdbb8404) will be amended with the rest of the changes.

Commit:
src: deploy: Handle command failure for statistics

In `kw deploy` feature, all commands tracked for statistics (except for
full deploy) don't handle failure for statistics purposes. The failure
of those commands is not handled at all. Either way, when a command
fails, e.g. `kw deploy --list` ssh connection fails because the target
remote machine is down, a status of `success` will be registered for
this statistic in the database (in the example,
`<ID>|list|success|<DATE>|<START_TIME>|<RUN_TIME>`).

This commit makes `kw deploy` take account of the failure of all those
commands when calling `statistics_manager`. We now capture the return
value of the effective command handlers (run_kernel_install,
modules_install, etc.) and use it to determine the `status` between
`success` and `failure`.

An interesting approach would be to pass the return value to
`statistics_manager` and let it handle it. This would move the if and
else statements from `deploy.sh` to `kwlib.sh` in the function
definition.

Also, it is worth pointing out that an `interrupted` command is not
added to the statistics and we should consider handling this case in
the future.

Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

davidbtadokoro commented:
**[UPDATE]**
This force push finishes the `build` and `deploy` (statistics) integration.

Commit:
src: pomodoro: Integrate with kw_db

The pomodoro feature used metadata files both for data reports and for
its functionalities (like showing registered tags).

To take advantage of `kw_db` and to lessen the responsibilities of the
pomodoro component (i.e. database management), this commit integrates
`kw pomodoro` feature with `kw_db`.

Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

davidbtadokoro commented:
**[UPDATE]**
Last force-push integrates `pomodoro`.

rodrigosiqueira commented on src/kw_db.sh:
```diff
@@ -152,8 +152,8 @@ function format_values_db()
     [[ "$count" -eq 0 ]] && values+='('
-    # check if not a sqlite function
-    if [[ ! "$val" =~ ^[[:alnum:]_]+\(.*\)$ ]]; then
+    # check if not a sqlite function nor NULL
```
Could you expand this function documentation to add one example about this 'correct format'?

rodrigosiqueira commented on database/kwdb.sql:
```diff
@@ -0,0 +1,222 @@
+-- Tables
+
+-- This table holds the metadata of the config files handled by kw
+CREATE TABLE IF NOT EXISTS "config" (
```
Rename it to `kernel_config`.

rodrigosiqueira commented on database/kwdb.sql:
```diff
@@ -0,0 +1,222 @@
+  "id" INTEGER NOT NULL UNIQUE,
+  "name" TEXT NOT NULL UNIQUE,
+  "description" TEXT,
+  "path" TEXT NOT NULL UNIQUE,
```
Maybe we add a field for tracking when the config file was saved.

rodrigosiqueira commented on database/kwdb.sql:
```diff
@@ -0,0 +1,222 @@
+CREATE TABLE IF NOT EXISTS "tag" (
+  "id" INTEGER NOT NULL UNIQUE,
+  "name" TEXT NOT NULL UNIQUE,
+  "active" INTEGER NOT NULL CHECK ("active" IN (0, 1)) DEFAULT 1,
+  PRIMARY KEY("id")
```
Could you add a comment that describes this field in more detail?

rodrigosiqueira commented on database/kwdb.sql:
```diff
@@ -0,0 +1,222 @@
+INSERT OR IGNORE INTO "command_label" ("name") VALUES
+  ('build'),
+  ('codestyle'),
+  ('configm'),
```
Update this command, it should be `kernel_config_manager`.

rodrigosiqueira commented on database/kwdb.sql:
```diff
@@ -0,0 +1,222 @@
+  ('maintainers'),
+  ('modules_deploy'),
+  ('ssh'),
+  ('uninstall');
```
Please, update this list with all the command options (e.g., VM, self-update, etc).

Commit:
src: report: Fix unreachable command

This commit fixes an unreachable command. This command was below a
return statement, which rendered it unreachable.

Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

Commit:
src: report: Integrate with kw_db

The `kw report` feature handles every type of report of kw tracked data.
As all data used to be stored using metadata files, `report` is
responsible for any logic related to fetching, processing, formatting,
and outputting reports. There are two points worth noting:
- The feature offers filters like by type of report (statistics or
  Pomodoro sessions) and periods of time (like a given day, week, month
  and year). The latter resulted in much of the feature's work being
  dealing with files hierarchy to fetch the proper data based on a
  period of time filter.
- Also, the file doesn't follow a pattern. For example, the logic for
  implementing the period of time filters for the Pomodoro sessions
  report is defined in `report.sh` whilst the one for the statistics
  report is defined in `statistics.sh`. The responsibilities of each
  function are not clear, as some fetch, process, and format the data
  (statistics case) and others just fetch and partly process the data
  (Pomodoro sessions case).

This commit solves the extensive logic implementation problem and lack
of standardized code by integrating `kw report` with `kw_db`.

A dedicated database and the power of SQL naturally solved the need for
complicated logic for applying period of time filters to data fetching.
This greatly reduced the complexity of the code and its size.

The integration also allowed us to enforce a pattern in the file. This
was done by:
- Unifying the function to fetch the raw data (`set_raw_data`);
- Adding a function to each type of report to process and format the
  data (`process_and_format_<statistics/pomodoro>_raw_data`);
- Unifying the output function (`show_report`).

Note:
- This change resulted in a revision of the `report` test suite that
  greatly decreased its running time.
- Just some processing functions like getting the maximum/minimum value
  from `statistics.sh` are still used. We could move them to `kwlib.sh`
  and deprecate `statistics.sh`.

Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

davidbtadokoro commented:
**[UPDATE]**
Last force-push integrates with `report`. Still need to make the changes requested above.

Commit:
src: kw_time_and_date: Move timebox_to_sec function

The same `timebox_to_sec` function is defined in the database migration
script and in `report.sh`. Also, this function possibly will be used in
`pomodoro.sh`.

As such, this commit moves this function to the library file
`kw_time_and_date.sh` where it can be included in the files listed
above. The commit also adds documentation, error handling, and automated
tests to the function.

Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

rodrigosiqueira commented on database/migrate_legacy_data_20220101.sh:
```diff
@@ -0,0 +1,226 @@
+#!/bin/bash
+
+# This file handles the migration of user data, from the old folders and files
```
/user data/of legacy user data/

rodrigosiqueira commented on database/migrate_legacy_data_20220101.sh:
```diff
@@ -0,0 +1,226 @@
+include "$KW_LIB_DIR/kwio.sh"
+include "$KW_LIB_DIR/kwlib.sh"
+include "$KW_LIB_DIR/kw_string.sh"
+include "$KW_LIB_DIR/kw_db.sh"
```
Please, review the code style in this entire file to ensure it follows kw standard.

rodrigosiqueira commented on database/migrate_legacy_data_20220101.sh:
```diff
@@ -0,0 +1,226 @@
+# This file handles the migration of user data, from the old folders and files
+# storage, to the new database model
+
+KW_LIB_DIR='src'
+KW_DB_DIR='database'
```
Let's make it readonly and move it after the include.

rodrigosiqueira commented on database/migrate_legacy_data_20220101.sh:
```diff
@@ -0,0 +1,226 @@
+  done
+}
+
+function db_migration()
```
How about `db_migration_main()`? Move it to the beginning of this file.

rodrigosiqueira commented on database/migrate_legacy_data_20220101.sh:
```diff
@@ -0,0 +1,226 @@
+
+function db_migration()
+{
+  execute_sql_script "$KW_DB_DIR/kwdb.sql"
```
Do you know if this command always passes? If not, we need to check its return.

rodrigosiqueira commented on database/migrate_legacy_data_20220101.sh:
```diff
@@ -0,0 +1,226 @@
+  local count=0
+  local -a values=()
+
+  file_list=$(find "${datadir}/statistics" -type f | sort)
```
Use `--dictionary-order` after `sort`.

rodrigosiqueira commented on database/migrate_legacy_data_20220101.sh:
```diff
@@ -0,0 +1,226 @@
+    while IFS=' ' read -r label_name elapsed_time_in_secs; do
+      status='success'
+      # The new database stores execution status, making this label unnecessary
+      if [[ "$label_name" == 'build_failure' ]]; then
```
We also need to handle `deploy_failure`.

rodrigosiqueira commented on database/migrate_legacy_data_20220101.sh:
```diff
@@ -0,0 +1,226 @@
+    done < "$file"
+  done
+
+  # insert_values '"statistics_report"' "$columns" 4 "${values[@]}"
```
We can drop it.

rodrigosiqueira commented on database/migrate_legacy_data_20220101.sh:
```diff
@@ -0,0 +1,226 @@
+    done < "$file"
+  done
+
+  # insert_values '"pomodoro_report"' "$columns" 5 "${values[@]}"
```
Drop this line.

rodrigosiqueira commented on database/migrate_legacy_data_20220101.sh:
```diff
@@ -0,0 +1,226 @@
+  local columns='("name","description","path")'
+  local count=0
+  local -a values=()
+
```
drop this extra line.

rodrigosiqueira commented on database/migrate_legacy_data_20220101.sh:
```diff
@@ -0,0 +1,226 @@
+  for file in $file_list; do
+    path="$file"
+    name="${path##*/}" # get just the file name
+    description=$(cat "${metadata_dir}/${name}")
```
Check: https://kworkflow.org/content/codingstyle.html#command-substitution-and-arithmetic-expression

rodrigosiqueira commented on database/migrate_legacy_data_20220101.sh:
```diff
@@ -0,0 +1,226 @@
+    fi
+  done
+
+  # insert_values '"config"' "$columns" 3 "${values[@]}"
```
Drop this line.

rodrigosiqueira commented on database/migrate_legacy_data_20220101.sh:
```diff
@@ -0,0 +1,226 @@
+  printf '%s\n' "$time_value"
+}
+
+function insert_values()
```
Can we delete this function?

rodrigosiqueira commented on setup.sh:
```diff
@@ -365,12 +367,20 @@ function synchronize_files()
   ASSERT_IF_NOT_EQ_ZERO "The command 'rsync -vr $CONFIG_DIR/ $etcdir $verbose' failed" "$?"
 
+  # Database files
+  mkdir -p "$databasedir"
+  cmd_output_manager "rsync -vr $DATABASE/ $databasedir" "$verbose"
```
Remember to use the complete parameter name.

rodrigosiqueira commented on setup.sh:
```diff
@@ -365,12 +367,20 @@ function synchronize_files()
   mkdir -p "$datadir"
   mkdir -p "$datadir/statistics"
   mkdir -p "$datadir/configs"
+  if [[ -x "$databasedir/migrate_legacy_data_20220101.sh" ]]; then
+    eval "$databasedir/migrate_legacy_data_20220101.sh"
+  fi
```
How will we set up the empty database if we don't have the migration script?

rodrigosiqueira commented on tests/kw_db_test.sh:
```diff
@@ -261,4 +261,121 @@ function test_select_from()
+  # valid operation with non-existent row
+  replace_into 'fake_table' '("name","attribute1","attribute2")' "('someName','someAtt1','someAtt2')"
+  ret="$?"
+  output=$(sqlite3 "$KW_DATA_DIR/kw.db" -batch "SELECT 'f'.'name' FROM 'fake_table' AS 'f' WHERE 'f'.'name'='someName' ;")
```
https://kworkflow.org/content/codingstyle.html#string-concatenation

check the other parts of this test.

rodrigosiqueira commented on src/kwlib.sh:
```diff
@@ -1,5 +1,6 @@
 # NOTE: src/kw_config_loader.sh must be included before this file
 include "$KW_LIB_DIR/kw_string.sh"
+include "$KW_LIB_DIR/kw_db.sh"
```
https://kworkflow.org/content/codingstyle.html#string-concatenation

rodrigosiqueira commented on src/kwlib.sh:
```diff
@@ -338,62 +339,35 @@ function detect_distro()
+  local start_date
+  local start_time
+  local row
+  local columns='("label_name","status","date","time","elapsed_time_in_secs")'
```
How about `database_columns`?

rodrigosiqueira commented on src/kwlib.sh:
```diff
@@ -338,62 +339,35 @@ function detect_distro()
+  start_date=$(date +%Y-%m-%d)
+  start_time=$(date +%H:%M:%S)
+
+  elapsed_time=$(date -d@"$elapsed_time_in_secs" -u +%H:%M:%S)
```
Replace `-u` with `--utc` and `-d` with `--date=`.

rodrigosiqueira commented on tests/kwlib_test.sh:
```diff
@@ -6,20 +6,19 @@ include './tests/utils.sh'
-  FAKE_STATISTICS_DAY_PATH="$FAKE_STATISTICS_MONTH_PATH/03"
-  ORIGINAL_DIR="$PWD"
+  declare -gr ORIGINAL_DIR="$PWD"
+  declare -gr FAKE_DATA="$SHUNIT_TMPDIR/db_testing"
```
https://kworkflow.org/content/codingstyle.html#string-concatenation

rodrigosiqueira commented on database/migrate_legacy_data_20220101.sh:
```diff
@@ -8,6 +8,7 @@ KW_DB_DIR='database'
 . 'src/kw_include.sh' --source-only
 include "$KW_LIB_DIR/kwio.sh"
 include "$KW_LIB_DIR/kwlib.sh"
+include "$KW_LIB_DIR/kw_time_and_date.sh"
```
https://kworkflow.org/content/codingstyle.html#string-concatenation

rodrigosiqueira commented on src/kw_time_and_date.sh:
```diff
@@ -1,3 +1,5 @@
+include "$KW_LIB_DIR/kw_string.sh"
```
https://kworkflow.org/content/codingstyle.html#string-concatenation

rodrigosiqueira commented on src/kw_db.sh:
```diff
@@ -209,7 +212,11 @@ function select_from()
   fi
 
-  sqlite3 -init "$KW_DB_DIR/pre_cmd.sql" -cmd "$pre_cmd" "$db_path" -batch "SELECT $columns FROM $table;"
+  query="SELECT $columns FROM $table ;"
+  if [[ -n "${order_by}" ]]; then
+    query="SELECT $columns FROM $table ORDER BY ${order_by} ;"
+  fi
+  sqlite3 -init "$KW_DB_DIR/pre_cmd.sql" -cmd "$pre_cmd" "$db_path" -batch "$query"
```
Can we just use `select_from`?

davidbtadokoro replied:
Sorry if I'm  missing something, but I don't think we can do this here as this is part of the definition of `select_from` itself.

davidbtadokoro commented on database/kwdb.sql:
```diff
@@ -0,0 +1,222 @@
+  "elapsed_time_in_secs"
+FROM
+  (SELECT "id" AS "e_id", "date", "time" FROM "event")
+JOIN "executed" ON "executed"."id" IS "e_id"
+JOIN "command_label" ON "command_label"."id" IS "executed"."command_label_id"
```
Sorry if I'm understanding it wrong. I think `JOIN` and `INNER JOIN` are equivalent. `NATURAL JOIN` requires the attributes in the `ON` clause to have the same name. The benefit is that, by doing this, the line would be less verbose and more informative for someone that knows what a `NATURAL JOIN` is (e.g. line 142 would be `NATURAL JOIN "executed"`. The downside is that it requires renaming of attributes in the file and doesn't explicitly indicates which attribute (or if there are more than one) is used in the join condition. Imo, I think keeping it this way is better. Any thoughts?

davidbtadokoro commented on tests/kernel_config_manager_test.sh:
```diff
@@ -96,26 +117,24 @@ function test_save_config_file_check_directories_creation()
-
-  assertTrue "${LINENO}: Failed to find ${NAME_1}" '[[ -f ${dot_configs_dir}/configs/${NAME_1} ]]'
-  assertTrue "${LINENO}: Failed the metadata related to ${NAME_1}" '[[ -f $dot_configs_dir/metadata/${NAME_1} ]]'
+  assertTrue "${LINENO}: Failed to find .config file for ${NAME_1}" '[[ -f ${dot_configs_dir}/${NAME_1} ]]'
```
Sorry, I didn't catch it.

davidbtadokoro commented on src/kw_db.sh:
```diff
@@ -95,10 +95,93 @@ function insert_into()
   sqlite3 -init "$KW_DB_DIR/pre_cmd.sql" "$db_path" -batch "INSERT INTO $table $entries VALUES $values;"
 }
 
+# This function updates or insert rows into table of given database,
```
Got it!

davidbtadokoro replied:
Done it!

Commit:
src: kw_db.sh: Move file to src/lib directory

As per the project structure, the `kw_db` is a library file and should
be placed in the `src/lib` directory. This commit does this by moving
`kw_db` from the `src` directory to the `src/lib` directory and making
the necessary adaptations throughout the project code.

Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

Commit:
src: pomodoro: Fix time limit of active timeboxes display

Active timeboxes for the feature `kw pomodoro --check-timer` have a time
limit of 86399 seconds (23:59:59 in H:M:S format) for both the elapsed
time and the remaining time. Because the function to convert time in
seconds representation to format H:M:S representation assumed a 24-Hour
Time Format (i.e. 00:00:00~23:59:59) a value of 72 hours (259200
seconds) will be displayed as 00:00:00.

This commit fixes this by using the function
`secs_to_arbitrarily_long_hours_min_secs` instead.

Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

davidbtadokoro commented on database/migrate_legacy_data_20220101.sh:
```diff
@@ -0,0 +1,232 @@
+  local -a values=()
+
+  # if there are no statistics reports to migrate, return
+  [[ ! -d "${datadir}/statistics" ]] && return 0
```
Mark migrated directories to not duplicate data.

Commit:
src: backup: Adapt to kw_db

When restoring a backup of kw data, if not using the `--force` flag, the
statistics and pomodoro reports, and the kernel configs data are
restored. This data is almost all the personal user data of kw (besides
the kw config files).

With the addition of `kw_db`, the file `KW_DATA_DIR/kw.db` needs to be
restored as well. Also, the legacy metadata files and directories don't
have the need to be restored and (in theory) should already be migrated
into `kw_db`, leaving just `KW_DATA_DIR/kw.db` and the actual kernel
config files to be restored.

This commit does this by not calling the functions that used to restore
the data mentioned in the first paragraph and adding and calling
functions that restore the `kw_db` file and actual kernel config files.

Note: As said before, this commit only affects the restoration without
`--force` as, in this case, the full decompressed backup is just copied
to `KW_DATA_DIR`.

Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

rodrigosiqueira commented:
I merged this change in the unstable branch.

Thanks a lot @kwy95 @davidbtadokoro @magalilemes for all the hard work. This is a fantastic change that put kw at another level :)
